File: chrwisewindow.py
#!/usr/bin/python
'''
File: chrwisewindow.py
Author: Rendong Yang
Description: the file is a modified version of chrwisewindw.pl for handling paired-end data
Email: cauyrd@gmail.com
Date: Mon Dec  5 14:31:37 EST 2011
'''
import sys

def get_position(pair1, pair2):
    position = [pair1[1],pair2[1],pair1[2],pair2[2]]
    position = map(int,position)
    position.sort()
    start = position[0]
    length = position[-1]-position[0]
    return (start,length)

def get_update_bin(bins, fragments, resolution):
    '''this function is used to calculate the bins coverage for each chromosome'''
    for i in range(len(fragments)):
        # get the bin index of start of each fragment
        (start_div,start_mod) = divmod(fragments[i][0],resolution)
        (fraglen_div,fraglen_mod) = divmod(fragments[i][1],resolution)
        if start_mod == 0:
            start_index = start_div - 1
            start_len = 1
        else:
            start_index = start_div 
            start_len = (resolution - start_mod)+1
            if fraglen_div == 0:
                if start_len > fraglen_mod:
                    bins[start_index] += fraglen_mod/float(resolution)
                    continue
                else:
                    bins[start_index] += start_len/float(resolution)
                    bins[start_index+1] += (fraglen_mod-start_len)/float(resolution)
                    continue
        bins[start_index] += start_len/float(resolution)
        full_cover_bin = (fragments[i][1] - start_len)/resolution
        for j in range(start_index+1,start_index+1+full_cover_bin):
            bins[j] += 1
        try:
            bins[j+1] += (fragments[i][1]-start_len)%resolution/float(resolution)
        except NameError:
            bins[start_index+1] += (fragments[i][1]-start_len)%resolution/float(resolution)
        except:
            print 'Error in fill_cover_bin computing!'
            sys.exit(0)
import re
import os

sp = sys.argv[1]
if sp.lower() == 'human' or 'hg19':
    nchr = 24
    fold1 = 'hg19'
elif sp.lower() == 'hg18':
	nchr = 24
	fold1 = 'hg18'
elif sp.lower() == 'mouse' or 'mm9':
    nchr = 21
    fold1 = 'mm9'
else:
    print 'species error!'
    exit(0)
infilename = sys.argv[2]
INFILE_pre = open(infilename)
infilename2 = INFILE_pre.readline().rstrip()
INFILE_pre.close()

INFILE = open(infilename2)
format = sys.argv[3]
readlenth = int(sys.argv[4])
resolution = int(sys.argv[5])
frontname = sys.argv[6]+'.windowhitscount.'
selectname = sys.argv[6]+'.select.'

format = sys.argv[3]
# get the all chromosome size of selected reference genome
tmp = len(sys.argv[0].split('/'))
path = '/'.join(sys.argv[0].split('/')[0:(tmp-1)])
CHROMOFILE = open(path+'data/'+fold1+'/chromoends.txt')
chromosize = [0]*nchr
for i,line in enumerate(CHROMOFILE):
    if sp.lower() == 'bed':
        chromosize[i] = int(line.rstrip()) - 1
    else:
        chromosize[i] = int(line.rstrip())

# get the fragments coverage from paired-end data
fragments = {}
chr = ['chr'+str(i) for i in range(1,23)] + ['chrX']+ ['chrY']
while True:
    pair1 = INFILE.readline().split()
    if not pair1:
        break
    pair2 = INFILE.readline().split()
    if pair1[0] != pair2[0] or pair1[0] not in chr:
        continue
    p = re.compile('chr(.*)')
    m = p.match(pair1[0])
    if m.group(1) == 'X':
        chr_index = nchr-2
    elif m.group(1) == 'Y':
        chr_index = nchr-1
    else:
        chr_index = int(m.group(1))-1
    if chr_index not in fragments:
        fragments[chr_index] = [get_position(pair1,pair2)]
    else:
        fragments[chr_index].append(get_position(pair1,pair2))

# output the window coverage for each chromosome in windowhitcount and select file
threshold =1 
for i in range(nchr):
    upperlimit = chromosize[i]/resolution
    if chromosize[i] % resolution == 0:
        chr_bins = [0]*upperlimit
    else:
        chr_bins = [0]*(upperlimit+1)
    #chr_bins = [0]*upperlimit if chromosize[i]%resolution==0 else [0]*(upperlimit+1)
    OUTFILE = open(frontname+'chr'+str(i+1)+'.txt','w')
    OUTSELECT = open(selectname+'chr'+str(i+1)+'.txt','w')
    if i in fragments.keys():
        get_update_bin(chr_bins,fragments[i],resolution)
    for j in range(upperlimit):
        start = resolution*j + 1
        end = start + resolution -1
        if (i == (nchr-2)):
            print >> OUTFILE, 'chrX\t'+str(start)+'\t'+str(end)+'\t'+str(chr_bins[j])
        elif (i == (nchr - 1)):
            print >> OUTFILE, 'chrY\t'+str(start)+'\t'+str(end)+'\t'+str(chr_bins[j])
        else:
            print >> OUTFILE, 'chr'+str(i+1)+'\t'+str(start)+'\t'+str(end)+'\t'+str(chr_bins[j])
        if chr_bins[j] > threshold:
            print >> OUTSELECT, str(j)+'\t'+str(chr_bins[j])

File: chrwisewindow.pl
#!/usr/bin/perl 
use POSIX;
use Cwd qw(realpath cwd);
use File::Basename;
my($fn, $dir) = fileparse(realpath($0));
# Steve Qin 05/15/08
# Changed by JC at 08/11/2008

$argc = @ARGV;
$argc == 8 || die "Provide input files names that include species info, multiple realign.txt filenames, file format, minimum and maximum DNA fragment length, read length, window size, and result filename.\n";

$threshold = 1;
#$readlength = 25;

$sp1 = $ARGV[0];
if ($sp1 eq "human" || "hg19") 
{
	$nchr = 24;
	$fold1 = "hg19";
}
elsif ($sp1 eq "hg18") 
{
	$nchr = 24;
	$fold1 = "hg18";
}
elsif ($sp1 eq "mouse" || "mm9") 
{
	$nchr = 21;
	$fold1 = "mm9";
}
else
{
        print "species error!\n";
        exit(0);
}

$innamename = $ARGV[1];
open(INNAMEFILE, $innamename); 
$format = $ARGV[2];
$fragmin = $ARGV[3];
$fragmax = $ARGV[4];
$fragmentWidth = $fragmax; #####($fragmin + $fragmax)/2;
$readlength = $ARGV[5];
$resolution = $ARGV[6];
$frontname = ">".$ARGV[7].".windowhitscount.";
$selectname = ">".$ARGV[7].".select.";
open(CHROMOFILE,"$dir/data/".$fold1."/chromoends.txt");

$count = 0;
my @filenames;
while($datline = <INNAMEFILE>)
{
        chomp($datline);
        if(length($datline) ==0)
        {
                next;
        }
	$filenames[$count] = $datline;
	$count ++;
}
$filecount = $count;
print "There are $filecount files.\n";
for($i = 0;$i < $nchr; $i++)
{
        $plushitscount[$i] = 0;
	$minushitscount[$i] = 0;
}
$totalcount = 0;
$count = 0;
for($i=0;$i<$filecount;$i++)
{
	$currentline = $filenames[$i];
	if ($currentline =~/(.*)\<format\>(.*)\<\/format\>/)
	{
		$tmpformat = lc($2);
		$tmpname = $1;
	}
	else 
	{
		$tmpformat = $format;
		$tmpname = $currentline;
	}
	&realignfileread($tmpname,$tmpformat,*totalcount, *count,*plushits, *minushits, *plushitscount, *minushitscount);
	print "total count = $totalcount.\n";
        print "$i\tcount = $count.\n";
}
$totalread = $count;

$count = 0;
my @chromosize;  
while($datline = <CHROMOFILE>)
{
        chomp($datline);
        @datas = split(" ",$datline);
	$chromosize[$count] = $datas[0];
	print "$chromosize[$count]\n";
	$count++;
	if($count == $nchr)
	{
		last;
	}
}
close(CHROMOFILE);

my @added;
$firstpart = $fragmin / $resolution;
$secondpart = ($fragmax - $fragmin) / $resolution;
if ($secondpart == 0)
{
	$totalparts = $fragmax / $resolution;
	$dimension = $totalparts + 1;
	for($j=0;$j<$resolution;$j++)
	{
		$startfrac = 1-$j/$resolution;
		$endfrac = 1-$startfrac;
		$added[$j * $dimension] = $startfrac ;
#		$added[$j * $dimension] = 1;
		for($k = 1;$k < $totalparts;$k++)
		{
			$added[$j * $dimension + $k] = 1;
		}
		$added[$j * $dimension + $totalparts] = $endfrac;
	}
#print "bobe\n@added\n";
#print "dimension $dimension, resolution $resolution $ARGV[5]\n";
#exit(0);
}
else
{
	$inv = 1/$secondpart;
	$totalparts = $fragmax / $resolution;
	$dimension = $totalparts + 1;
	for($j=0;$j<$resolution;$j++) 
	{
		$startfrac = 1-$j/$resolution;
        	$endfrac = 1-$startfrac;         
		$added[$j * $dimension] = $startfrac ;
#		$added[$j * $dimension] = 1;
        	for($k = 1;$k < $firstpart;$k++)
        	{
                	$added[$j * $dimension + $k] = 1;
        	}
		$added[$j * $dimension + $firstpart] = $endfrac + $startfrac * (1 - 0.5 * $startfrac * $inv);        
		for($k = $firstpart + 1;$k < $totalparts;$k++)
        	{
                	$added[$j * $dimension + $k] = 1 - ($k - $firstpart - 0.5 + $startfrac) * $inv;
        	}
		$added[$j * $dimension + $totalparts] = 0.5 * $endfrac * $inv *$endfrac;
	}
#print "bobe\n@added\n";
#exit(0);
}

for($i = 0;$i < $nchr; $i++)
{
	$chromosome = $i +1;
	open(OUTFILE, $frontname."chr".$chromosome.".txt");	
        open(OUTSELECTFILE, $selectname."chr".$chromosome.".txt");
	my @pluspos;
	for($j=0;$j<$plushitscount[$i];$j++)
	{
		$pluspos[$j] = $plushits[$i][$j];
	}
        @sortpos = sort {$a <=> $b} @pluspos;
	$plustotal = $plushitscount[$i];

	$upperlimit = floor($chromosize[$i]/$resolution) + 1;
	for($j = 0;$j < $upperlimit; $j++)
        {
		$heights[$j] = 0;		
	}
	print "chr ", $i+1, " upperlimit = $upperlimit.\n";
	for($j = 0;$j < $plustotal; $j++)
        {
		$lowend = floor(($sortpos[$j]-1) /$resolution);##### 03/26/08 ( -1)
                $residual = $sortpos[$j] - 1 - $lowend *$resolution;
        	if(($lowend >= 0)&&($lowend < ($upperlimit - $totalparts)))
		{
			for($k=0;$k<($totalparts+1);$k++)
        	        {
				$heights[$lowend + $k] = $heights[$lowend + $k] + $added[$residual * $dimension + $k];
#				print "heights ($lowend + $k), ($residual * $dimension + $k),$added[$residual * $dimension + $k]\n";
			}
		}
#		exit(0);
	}
#	print "pos: $sortpos[1]\n";
#	exit(0);
	my @minuspos;
        for($j=0;$j<$minushitscount[$i];$j++)
        {
                $minuspos[$j] = $minushits[$i][$j];
        }
        @sortpos = sort {$a <=> $b} @minuspos;
        $minustotal = $minushitscount[$i];
	for($j = 0;$j < $minustotal; $j++)
        {
                $lowend = floor(($sortpos[$j]-1) /$resolution);##### 03/26/08 ( -1)
                $residual = $sortpos[$j] - 1 - $lowend *$resolution;
		$residual = $resolution -1 - $residual;
                if(($lowend >= 0)&&($lowend < ($upperlimit - $totalparts)))
                {
                        for($k=$totalparts;$k>=0;$k--)
                        {
                                $heights[$lowend + $k] = $heights[$lowend + $k] + $added[$residual * $dimension + ($totalparts - $k)];
			}
                }
        }
#	print "minus: $sortpos[1]\n";
#	exit(0);
	$selected = 0;
        for($j=0;$j<$upperlimit;$j++) 
        { 
		$start = $resolution*$j + 1;
		$end = $start + $resolution -1;
		if($i == ($nchr - 2))
		{
			print OUTFILE "chrX\t",$start,"\t", $end, "\t",$heights[$j],"\n";
		}
		elsif($i == ($nchr-1))
		{
			print OUTFILE "chrY\t",$start,"\t", $end, "\t",$heights[$j],"\n";
		}
		else
		{
			print OUTFILE "chr", $i + 1,"\t",$start,"\t", $end, "\t",$heights[$j],"\n";	
		}
                if($heights[$j]>$threshold)
                {
                        print OUTSELECTFILE "$j\t$heights[$j]\n";
                	$selected ++;
		}
	} 
	print "chr ", $i+1," # nonempty windows = ",$selected,"\n";
	close(OUTFILE);
	close(OUTSELECTFILE);
}

print "total number of reads = $totalcount.\n";
print "total number of successfully aligned reads = $totalread. (",100*$totalread/$totalcount," %)\n";

sub realignfileread()
{
	local ($subcurrentfilename, $subformat, *subtotalcount, *subcount, *subplushits, *subminushits, *subplushitscount, *subminushitscount) = @_; 
	if($subcurrentfilename =~ /gz$/)
	{
		system("gunzip ".$subcurrentfilename);
		$namelength = length($subcurrentfilename);		
		$newname = substr($subcurrentfilename,0,$namelength-3); 
		print "name = $newname\n";
                open(INREADFILE, $newname);
	}
	elsif($subcurrentfilename =~ /zip$/)
        {
                system("unzip ".$subcurrentfilename);
		$newname = substr($subcurrentfilename,0,$namelength-4);
                open(INREADFILE, $newname.".txt");
	}
	else
	{
		open(INREADFILE, $subcurrentfilename);		
	}
	while($datline = <INREADFILE>)
	{
		my $chr=$posit=$direct=$loc=0;
		chomp($datline);
		my @datas=split /[\s\t]+/, $datline;
		$subtotalcount ++;
		if($subformat eq "eland")
		{
			if(($datline=~ /chr/)&&($datline =~/\.fa/))
			{
				$posit = $datas[7];
				$direct = $datas[8];
				$loc=$datas[6];
			}
		}
		elsif ($subformat=~/custom\[(.*)\]/)
		{
			$tempo = $1;
			if(($datline=~ /chr/)&&($datline =~/\.fa/))
			{
				my @infos=split /\,\s*/,$tempo;
				$loc=$datas[$infos[0]-1];
				$posit=$datas[$infos[1]-1];
				$direct=$datas[$infos[2]-1];
			}
		}
		elsif ($subformat eq "bed") 
		{
			$chr =$2 if($datas[0]=~/^(chr|c)?(\d+|X|Y)$/);
			$posit = $datas[1]+1;
			$direct = $datas[-1];
		}
		$chr=$1 if($loc=~/chr(\d+|X|Y)\.fa/ || $loc=~/chromosome\.(\d+|X|Y)\.fa/);
		next if(!$chr);
		$chr=($nchr-1) if($chr eq "X");
		$chr=$nchr if($chr eq "Y");
		if($direct eq "+" || $direct eq "F")
		{
			$subplushits[$chr-1][$subplushitscount[$chr-1]] = $posit;
#			print "$posit \n";
#			exit(0);
			$subplushitscount[$chr-1] ++;
		}
		elsif($direct eq "-" || $direct eq "R")
		{
			$subminushits[$chr-1][$subminushitscount[$chr-1]] = $posit + $readlength - $fragmentWidth;
#			print "$posit + $readlength - 1\n";
#			exit(0);
			$subminushitscount[$chr-1] ++;
		}
		else
		{
			print "There is strand error. strand is $direct, position is $posit...............\n\n\n";
			exit(0);
		}
		$subcount++;
	}
close(INREADFILE);
}

